src/mpmc_relaxed_queue.ml

```diff
@@ -0,0 +1,269 @@
+  array). In fact, it could be any number. Thus, all operations on slot have to use 
+  CAS to ensure that no item is overwrriten on store and no item is dequeued by two 
+  threads at once. 
+
+  Above works okay in practise, and there is some relevant literature, e.g.  
+  (DOI: 10.1145/3437801.3441583) analyzed this particular design. There's also 
+  plenty older papers looking at similar approaches 
+  (e.g. DOI: 10.1145/2851141.2851168). 
+
+  Note, this design may violate FIFO (on overlap). The risk can be minimized by 
+  ensuring size of array >> number of threads but it's never zero.
+  (github.com/rigtorp/MPMCQueue has a nice way of fixing this, we could add it).
+
+  # Non-blocking operations
+
+  Up until now [push] and [pop] were allowed to block indefinitely on empty and full 
+  queue. Overall, what can be done in those states?
+
+  1. Busy wait until able to finish.
+  2. Rollback own index with CAS (unassign itself from slot).
+  3. Move forward other index with CAS (assign itself to the same slot as opposite 
+  action).
+  4. Mark slot as burned - dequeue only.
+
+  Which one then?
+
+  Let's optimize for stability, i.e. some reasonable latency that won't get much worse
+  under heavy load. Busy wait is great because it does not cause any contention in the
+  hotspots ([head], [tail]). Thus, start with busy wait (1). If queue is busy and 
+  moving fast, there is a fair chance that within, say, 30 spins, we'll manage to 
+  complete action without having to add contention elsewhere. 
+  
+  Once N busy-loops happen and nothing changes, we probably want to return even if its 
+  costs. (2), (3) both allow that. (2) doesn't add contention to the other index like 
+  (3) does. Say, there's a lot more dequeuers than enqueuers, if all dequeurs did (3), 
+  they would add a fair amount of contention to the [tail] index and slow the 
+  already-outnumbered enqueuers further. So, (2) > (3) for that reason.
+
+  However, with just (2), some dequeuers will struggle to return. If many dequeuers 
+  constatly try to pop an element and fail, they will form a chain.
+
+   tl                 hd
+   |                  |
+  [.]-[A]-[B]-[C]-..-[X]
+
+  For A to rollback, B has to rollback first. For B to rollback C has to rollback first.
+
+  [A] is likely to experience a large latency spike. In such a case, it is easier for [A]
+  to do (3) rather than hope all other active dequeuers will unblock it at some point. 
+  Thus, it's worthwile also trying to do (3) periodically.
+
+  Thus, the current policy does (1) for a bit, then (1), (2) with periodic (3). 
+
+  What about burned slots (4)?
+
+  It's present in the literature. Weakly I'm not a fan. If dequeuers are faster to remove 
+  items than enqueuers supply them, slots burned by dequeuers are going to make enqueuers 
+  do even more work.
+
+  # Resizing 
+
+  The queue does not support resizing, but it can be simulated by wrapping it in a 
+  lockfree list. 
+*)
```

Comment:
Is this documentation up to date with the last changes ?

Reply:
Yeah. Added a couple corrections just now to refer to non-lf paths.